test/distributed/test_c10d.py

```diff
@@ -3048,32 +3052,86 @@ def fut_then(fut):
+        """
+        This unit test verifies whether the Future object is passed properly using gloo backend.
+        The hook callback function creates a Future object and sets a value to it.
+        """
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.world_size)
 
-        # Run backward
-        output.mean().backward()
+        # Get GPU model with simple_hook registered.
+        gpu_model = self._gpu_model_with_ddp_comm_hook(process_group, self._simple_hook)
 
-        # check whether the grads are equal to what then callback returns.
+        # check whether the grads are equal to what simple_hook's then callback returns.
         # without the comm_hook, result would be 0.25 * torch.ones(2, 2).
-        [self.assertEqual(p.grad, 2 * torch.ones(2, 2)) for p in model.parameters()]
+        self._run_and_verify_hook(gpu_model, 8 , 2 * torch.ones(2, 2))
 
-    @requires_gloo()
+    @requires_nccl()
     @skip_if_lt_x_gpu(2)
-    def test_ddp_comm_hook_future_passing_gpu_gloo(self):
+    def test_ddp_comm_hook_future_passing_gpu_nccl(self):
         """
-        This unit test executes _test_ddp_comm_hook_future_passing_gpu using gloo backend.
+        This unit test verifies whether the Future object is passed properly using nccl backend.
+        The hook callback function creates a Future object and sets a value to it.
         """
-        self._test_ddp_comm_hook_future_passing_gpu(c10d.ProcessGroupGloo)
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        # Get GPU model with simple_hook registered.
+        gpu_model = self._gpu_model_with_ddp_comm_hook(process_group, self._simple_hook)
+
+        # check whether the grads are equal to what simple_hook's then callback returns.
+        # without the comm_hook, result would be 0.25 * torch.ones(2, 2).
+        self._run_and_verify_hook(gpu_model, 8 , 2 * torch.ones(2, 2))
 
     @requires_nccl()
     @skip_if_lt_x_gpu(2)
-    def test_ddp_comm_hook_future_passing_gpu_nccl(self):
+    def test_ddp_comm_hook_allreduce_hook_nccl(self):
         """
-        This unit test executes _test_ddp_comm_hook_future_passing_gpu using nccl backend.
+        This unit test verifies whether a DDP communication hook that just calls
+        allreduce gives the same result result with the case of no hook registered.
+        Without the then callback, the future_value in reducer is no longer
+        a PyObject, and this unit test verifies future_value is properly checked.
         """
-        self._test_ddp_comm_hook_future_passing_gpu(c10d.ProcessGroupNCCL)
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        def allreduce_hook(state: object, bucket: dist.GradBucket) -> torch.futures.Future:
```

Comment:
I see, maybe both `torch.futures.Future` and `.get_future()` will return `torch.jit.Future`. Can we modify the return type to `torch._C.Future` for now, and we can change it back to `torch.futures.Future` when it really behaves the same way.